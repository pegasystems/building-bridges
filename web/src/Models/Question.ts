import { camelizeKeys } from 'humps';
import { UserVote } from ".";

export default class QuestionModel {
    _id!: string;
    content!: string;
    isAuthor!: boolean;
    hideVotes!: boolean;
    isAnonymous!: boolean;
    authorFullName: string = '';
    authorEmail: string = '';
    authorNickname: string = '';
    upvotes: number = 0;
    downvotes: number = 0;
    voted: UserVote = UserVote.None;
    read: string = 'false';
    hidden: boolean = false;

    addUserVote(voteType: UserVote) {
        if (voteType === UserVote.Up) {
            this.upvotes++;
            this.downvotes -= Number(this.voted !== UserVote.None);
            this.voted = UserVote.Up;
        } else {
            this.downvotes++;
            this.upvotes -= Number(this.voted !== UserVote.None);
            this.voted = UserVote.Down;
        }
    }

    removeUserVote(): void {
        if (this.voted === UserVote.Up) {
            this.upvotes--;
        }
        else {
            this.downvotes--;
        }
    }

    toggleRead(): void {
        this.read = this.read === 'true' ? 'false' : 'true'
    }
    
    static createQuestionFromApiResult(questionDict: any): QuestionModel {
        let question = new QuestionModel();
        for (const [key, value] of Object.entries(camelizeKeys(questionDict, function (key, convert) {
            return key === '_id' ? key : convert(key);
        }))) {
            (question as any)[key] = value
        }

        question.hideVotes = question.upvotes == null && question.read === 'false';
        return question;
    }
}
