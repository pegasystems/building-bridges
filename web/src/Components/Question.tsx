import React from 'react';
import * as Models from '../Models'
import {SURVEYS_API, QUESTIONS_ENDPOINT} from '../Consts'
import Eye from './Eye';


export enum UserVote {
    Up = "up",
    Down = "down",
    None = "none",
}


export interface QuestionProps { 
    surveyKey: string;
    question: Models.Question;
    adminSecret: string | undefined;
    deleteQuestionCallback(question_id: string): void;
    addVoteCallback(question: Models.Question, voteType: Models.UserVote): any;
    deleteVoteCallback(question: Models.Question): any;
    markAsReadCallback(question: Models.Question): any;
    updateQuestionInState(question: Models.Question): any;
}

export default class Question extends React.Component<QuestionProps, {}> {

    deleteQuestion = (e: React.FormEvent<HTMLFormElement>) => {
        fetch(`${SURVEYS_API}${this.props.surveyKey}${QUESTIONS_ENDPOINT}/${this.props.question._id}`, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                this.props.deleteQuestionCallback(this.props.question._id);
            }
        });
        e.preventDefault();
    };

    newVote(question: Models.Question, newVoteType: Models.UserVote)  {
        fetch(`${SURVEYS_API}${this.props.surveyKey}${QUESTIONS_ENDPOINT}/${question._id}/vote?type=${newVoteType}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                this.props.addVoteCallback(question, newVoteType);
            }
        });
    }

    deleteVote(question: Models.Question) {
        fetch(`${SURVEYS_API}${this.props.surveyKey}${QUESTIONS_ENDPOINT}/${question._id}/vote`, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                this.props.deleteVoteCallback(question);
            }
        });
    }

    vote = (question: Models.Question, newVoteType: Models.UserVote, e: React.FormEvent<HTMLFormElement>) => {
        if (question.voted === newVoteType) {
            this.deleteVote(question);
        }
        else {
            this.newVote(question, newVoteType);
        }
        e.preventDefault();
    }

    markAsRead = (question: Models.Question, e: React.FormEvent<HTMLFormElement>) => {

        fetch(`${SURVEYS_API}${this.props.surveyKey}${QUESTIONS_ENDPOINT}/${question._id}/ctx?read=${question.read === 'true' ? 'false' : 'true'}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {     
                this.props.markAsReadCallback(Models.Question.createQuestionFromApiResult(question));
            }
        });

        e.preventDefault();
    }

    markAsHidden = () => {
        fetch(`${SURVEYS_API}${this.props.surveyKey}${QUESTIONS_ENDPOINT}/${this.props.question._id}?admin_secret=${this.props.adminSecret}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({hidden: !this.props.question.hidden})
        })
        this.props.question.hidden = !this.props.question.hidden;
        this.props.updateQuestionInState(this.props.question);
    };
    
    render(): JSX.Element {
        const question = this.props.question;
        const deleteButton = question.isAuthor && !this.props.adminSecret && (question.upvotes + question.downvotes === 0 || question.hideVotes) ?
            <form onSubmit={this.deleteQuestion}>
            <input type="hidden" name="id" value={question._id}/>
            <input type="submit" value="x" title="Remove this question" className="remove"/>
            </form> :
            <div/>;
        const hideButton = this.props.adminSecret ?
            <Eye slashed={this.props.question.hidden} hideOrShowCallback={this.markAsHidden}></Eye>
            : <div/>;
        const questionReadMarker = question.read === 'true' ? 'read' : ''
        const userVote = question.voted !== 'none' ? question.voted : questionReadMarker
        return (<li className={'vote ' + (this.props.adminSecret ? "" : userVote) + (this.props.question.hidden ? 'hidden' : '')} id={question._id}>
            <div className="right">
                <div className="thumbs">
                    <div className="item">
                        <form onSubmit={e => this.vote(question, Models.UserVote.Up, e)}>
                            <input type="hidden" name="id" value={question._id}/>
                            <input className="submit-up" type="submit" value={!question.hideVotes ? question.upvotes : ''}/>
                        </form>
                    </div>
                    <div className="item">
                        <form onSubmit={e => this.vote(question, Models.UserVote.Down, e)}>
                            <input type="hidden" name="id" value={question._id}/>
                            <input className="submit-down" type="submit" value={!question.hideVotes ? question.downvotes : ''}/>
                        </form>
                    </div>
                    <div className={'mark-as-read item ' + (question.read === 'true' ? 'read' : '')}>
                        <form onSubmit={e => this.markAsRead(question, e)}>
                            <input type="hidden" name="id" value={question._id}/>
                            <input title={'Mark as ' + (question.read === 'true' ? 'unread' : 'read')}
                                   className="submit-read" type="submit" value=""/>
                        </form>
                    </div>
                </div>
            </div>
            <div className="center">{question.content}</div>
            <a href={`#${question._id}`} title="Direct link to this post" className="permalink">#</a>
            {deleteButton}
            {hideButton}
        </li>);
    }
}
