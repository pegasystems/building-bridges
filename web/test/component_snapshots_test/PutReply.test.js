// Link.react.test.js
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Question from '../../src/Components/Question';

Enzyme.configure({ adapter: new Adapter() });

global.flushPromises = () => new Promise(resolve => setImmediate(resolve));

var sampleQuestionComponent = <Question
    question={{
        "_id": "id", "isAnonymous": true, "reply": "sample reply"
    }}
    adminSecret="secret"
/>


test('Reply button renders properly test', () => {
    const component = renderer.create(sampleQuestionComponent);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});



test('Reply workflow test', () => {
    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({}),
            ok: true
        })
    );

    const wrapper = mount(sampleQuestionComponent);
    const expectedHtmlBeforeClick = '<div>sample reply</div>'
    const expectedHtmlAfterClick = '<div><form><textarea id="responseBox" class="response" name="question">sample reply</textarea><label for="responseBox"></label><input class="add-new-button" type="submit" value="Submit"></form></div>'
    expect(wrapper.html().includes(expectedHtmlBeforeClick)).toEqual(true);
    wrapper.find('[className="svg-inline--fa fa-reply fa-w-16 reply-button"]').simulate('click')
    expect(wrapper.html().includes(expectedHtmlAfterClick)).toEqual(true);
    wrapper.find('input.add-new-button').simulate('submit')

    return flushPromises().then(() => {
        expect(wrapper.html().includes(expectedHtmlBeforeClick)).toEqual(true);
        expect(wrapper.html().includes(expectedHtmlAfterClick)).toEqual(false);
    });

});




test('Reply workflow with error test', () => {

    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({ 'message': 'wrong' }),
            ok: false
        })
    );
    const wrapper = mount(sampleQuestionComponent);
    wrapper.find('[className="svg-inline--fa fa-reply fa-w-16 reply-button"]').simulate('click')
    wrapper.find('input.add-new-button').simulate('submit')

    return flushPromises().then(() => {
        expect(wrapper.html().includes('wrong')).toEqual(true);
    });

});