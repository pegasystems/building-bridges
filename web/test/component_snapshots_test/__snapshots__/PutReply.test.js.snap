// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reply button renders properly test 1`] = `
<div>
  <li
    className="vote "
    id="id"
  >
    <div
      className="right"
    >
      <div
        className="thumbs"
      >
        <div
          className="item"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              name="id"
              type="hidden"
              value="id"
            />
            <input
              className="submit-up"
              type="submit"
            />
          </form>
        </div>
        <div
          className="item"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              name="id"
              type="hidden"
              value="id"
            />
            <input
              className="submit-down"
              type="submit"
            />
          </form>
        </div>
        <div
          className="mark-as-read item "
        >
          <form
            onSubmit={[Function]}
          >
            <input
              name="id"
              type="hidden"
              value="id"
            />
            <input
              className="submit-read"
              title="Mark as read"
              type="submit"
              value=""
            />
          </form>
        </div>
      </div>
    </div>
    <div
      className="center"
    >
      <div />
      <div
        className="footnote"
      />
    </div>
    <a
      className="permalink"
      href="#id"
      title="Direct link to this post"
    >
      #
    </a>
    <div />
    <i
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-eye fa-w-18 hide-eye"
        data-icon="eye"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </i>
    <i
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-reply fa-w-16 reply-button"
        data-icon="reply"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.309 189.836L184.313 37.851C199.719 24.546 224 35.347 224 56.015v80.053c160.629 1.839 288 34.032 288 186.258 0 61.441-39.581 122.309-83.333 154.132-13.653 9.931-33.111-2.533-28.077-18.631 45.344-145.012-21.507-183.51-176.59-185.742V360c0 20.7-24.3 31.453-39.687 18.164l-176.004-152c-11.071-9.562-11.086-26.753 0-36.328z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </i>
  </li>
  <div
    className="response"
  >
    <div>
      sample reply
    </div>
    <div
      className="footnote"
    >
      Answered by survey admin
    </div>
  </div>
</div>
`;
